gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -187,82 +193,82 @@ public void startUpload(InputStream pipeSource) {
+      if (writeOffset > 0) {
+        writeOffset = getCommittedWriteSize(uploadId);
+      }
+      responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset);
+      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
+      StreamObserver<InsertObjectRequest> requestStreamObserver =
+          stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), MILLISECONDS)
+              .insertObject(responseObserver);
+
+      // Wait for streaming RPC to become ready for upload.
+      try {
+        responseObserver.ready.await();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(
+            String.format(
+                "Streaming RPC failed to become ready for resumable upload for '%s'",
+                getResourceString()),
+            e);
+      }
 
-        try {
-          responseObserver.done.await();
-          if (responseObserver.hasTransientError()) {
-            continue;
+      boolean objectFinalized = false;
+      while (!objectFinalized) {
+        InsertObjectRequest insertRequest;
+        if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+          insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
+          writeOffset += insertRequest.getChecksummedData().getContent().size();
+        } else {
+          ByteString data =
+              ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
+          dataChunkMap.put(writeOffset, data);
+          if (dataChunkMap.size() >= NUMBER_OF_REQUESTS_TO_RETAIN) {
+            dataChunkMap.remove(dataChunkMap.firstKey());
           }
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format("Resumable upload failed for '%s'", getResourceString()), e);
+          insertRequest = buildInsertRequest(writeOffset, data, false);
+          writeOffset += data.size();
         }
+        requestStreamObserver.onNext(insertRequest);
+        objectFinalized = insertRequest.getFinishWrite();
+
+        if (responseObserver.hasTransientError()
+            || responseObserver.hasNonTransientError()) {
+          requestStreamObserver.onError(
+              responseObserver.hasTransientError()
+                  ? responseObserver.transientError
+                  : responseObserver.nonTransientError);
+          break;
+        } else if (objectFinalized) {
+          requestStreamObserver.onCompleted();
+        }
+      }
 
-        return responseObserver.getResponseOrThrow();
+      try {
+        responseObserver.done.await();
+        if (responseObserver.hasTransientError()) {
+          throw new IOException(responseObserver.transientError);
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
```

Comment:
Should it be something like, to preserve original failure and resource:
```
throw new RuntimeException(
    String.format("Failed to start resumable upload for '%s'", getResourceString()), e);
```

Reply:
All non-IOExceptions will be catched by the try-catch around retryer and the original failure and resource will be preserved.

Reply:
In this case it could be better to not catch `InterruptedException` here and special case it in `catch` around `retry` call?

Reply:
I think this works for me, updated.